craft_providers/actions/snap_installer.py

```diff
@@ -0,0 +1,202 @@
+def _get_host_snap(snap_name: str) -> Iterator[pathlib.Path]:
+    """Get snap installed on host.
+
+    Snapd provides an API to fetch a snap. First use that to fetch a snap.
+    If the snap is installed using `snap try`, it may fail to download. In
+    that case, attempt to construct the snap by packing it ourselves.
+
+    :yields: Path to snap which will be cleaned up afterwards.
+    """
+    with _temporary_file(f"{snap_name}.snap") as snap_path:
+        try:
+            _download_host_snap(snap_name=snap_name, output=snap_path)
+        except SnapInstallationError:
+            logger.warning(
+                "Failed to fetch snap from snapd, falling back to `snap pack` to recreate."
+            )
+            _pack_host_snap(snap_name=snap_name, output=snap_path)
+
+        yield snap_path
+
+
+def inject_from_host(*, executor: Executor, snap_name: str) -> None:
+    """Inject snap from host snap.
+
+    :raises SnapInstallationError: on unexpected error.
+    """
+    target_snap_path = f"/tmp/{snap_name}.snap"
+
+    try:
+        # Clean outdated snap, if exists.
+        executor.execute_run(
+            ["rm", "-f", target_snap_path],
+            check=True,
+            capture_output=True,
+        )
+
+        with _get_host_snap(snap_name) as host_snap_path:
+            executor.push_file(
+                source=host_snap_path,
+                destination=pathlib.Path(target_snap_path),
+            )
+
+        executor.execute_run(
+            ["snap", "install", "--dangerous", "--classic", target_snap_path],
+            check=True,
+            capture_output=True,
+        )
+    except subprocess.CalledProcessError as error:
+        raise SnapInstallationError(
+            brief=f"Failed to inject snap {snap_name!r}.",
+            details=details_from_called_process_error(error),
+        ) from error
+    except ProviderError as error:
+        raise SnapInstallationError(
+            brief=f"Failed to inject snap {snap_name!r}.",
+        ) from error
+
+
+def install_from_store(
+    *,
+    executor: Executor,
+    snap_name: str,
+    channel: str,
+    classic: bool,
+) -> None:
+    """Install snap from store into target.
+
+    Perform installation using method which prevents refreshing.
+
+    :param executor: Executor for target.
+    :param snap_name: Name of snap to install.
+    :param channel: Channel to install from.
+    :param classic: Install in classic mode.
+
+    :raises SnapInstallationError: on unexpected error.
+    """
+    try:
+        executor.execute_run(
+            [
+                "snap",
+                "download",
+                snap_name,
+                f"--channel={channel}",
+                f"--basename={snap_name}",
+                "--target-directory=/tmp",
+            ],
+            check=True,
+            capture_output=True,
+        )
+
+        install_cmd = ["snap", "install", f"/tmp/{snap_name}.snap", "--dangerous"]
+        if classic:
```

Comment:
We always use `--classic` when injecting and it's ignored when the snap is strictly confined, why can't we do the same here?

Reply:
Ah, good point.  Didn't realize that :D

Reply:
Fixed

Reply:
it used be an error

Reply:
Yeah I was surprised it didn't cause an issue with the integration tests